File: MatrixChainMultiplication/MatrixChainMultiplication.py

def bf_recursion(matrices, i, j):
    if i == j:
        return 0
    values = []
    for k in range(i, j):
        values.append(
            bf_recursion(matrices, i, k) +
            bf_recursion(matrices, k + 1, j) +
            (matrices[i][0] * matrices[k + 1][0] * matrices[j][1])
        )
    return min(values)


def brute_force(file_path):
    matrices = [tuple(int(i) for i in a.split(', ')) for a in open(file_path).read()[2:-2].split('}, {')]
    return bf_recursion(matrices, 0, len(matrices) - 1)


def dp_recursion(matrices, i, j, cache):
    if i == j:
        return 0
    if cache[i][j]:
        return cache[i][j]
    values = []
    for k in range(i, j):
        values.append(
            dp_recursion(matrices, i, k, cache) +
            dp_recursion(matrices, k + 1, j, cache) +
            (matrices[i][0] * matrices[k + 1][0] * matrices[j][1])
        )
    cache[i][j] = min(values)
    return cache[i][j]


def dynamic_programming_recursion(file_path):
    matrices = [tuple(int(i) for i in a.split(', ')) for a in open(file_path).read()[2:-2].split('}, {')]
    cache = [[0 for _ in range(len(matrices))] for _ in range(len(matrices))]
    return dp_recursion(matrices, 0, len(matrices) - 1, cache)


def dynamic_programming(file_path):
    matrices = [tuple(int(i) for i in a.split(', ')) for a in open(file_path).read()[2:-2].split('}, {')]
    cache = [[0 for _ in range(len(matrices))] for _ in range(len(matrices))]
    for a in range(1, len(matrices)):
        for i in range(len(matrices) - a):
            j = a + i
            values = []
            for k in range(i, j):
                values.append(cache[i][k] + cache[k + 1][j] + (matrices[i][0] * matrices[k + 1][0] * matrices[j][1]))
            cache[i][j] = min(values)
    return cache[0][len(matrices) - 1]

File: MatrixChainMultiplication/MatrixChainMultiplication.cpp
#include <string>
#include <fstream>
#include <sstream>
#include <algorithm>
#include "MatrixChainMultiplication.h"
using namespace std;

struct MatrixChainMultiplication::MatrixSize {
    int x;
    int y;
};

vector<MatrixChainMultiplication::MatrixSize> MatrixChainMultiplication::parseInput(const string& filePath) {
    vector<MatrixSize> matrices;
    ifstream file(filePath);
    string input;
    if (file.is_open()) {
        while (getline(file, input, '{')) {
            if (!input.empty()) {
                string pair;
                stringstream ss(input);
                MatrixSize matrixSize{};
                getline(ss, pair, ' ');
                matrixSize.x = stoi(pair.substr(0, pair.size() - 1));
                getline(ss, pair, ' ');
                matrixSize.y = stoi(pair.substr(0, pair.size() - 2));
                matrices.push_back(matrixSize);
            }
        }
        file.close();
    }
    return matrices;
}

int MatrixChainMultiplication::bfRecursion(MatrixSize* matrices, int i, int j) {
    if (i == j) {
        return 0;
    }
    int* values = new int[j - i];
    for (int k = 0; k < j - i; k++) {
        int m = matrices[i].x * matrices[k + i + 1].x * matrices[j].y;
        values[k] = bfRecursion(matrices, i, k + i) + bfRecursion(matrices, k + i + 1, j) + m;
    }
    int min = *min_element(values, values + (j - i));
    delete [] values;
    return min;
}

int MatrixChainMultiplication::bruteForce(const string& filePath) {
    vector<MatrixSize> matricesInput = parseInput(filePath);
    int matricesSize = int(matricesInput.size());
    MatrixSize matrices[matricesSize];
    for (int i = 0; i < matricesSize; i++) {
        matrices[i] = matricesInput[i];
    }
    return bfRecursion(matrices, 0, matricesSize - 1);
}

int MatrixChainMultiplication::dpRecursion(MatrixSize *matrices, int i, int j, int **cache) {
    if (i == j) {
        return 0;
    }
    if (cache[i][j] != 0) {
        return cache[i][j];
    }
    int* values = new int[j - i];
    for (int k = 0; k < j - i; k++) {
        int m = matrices[i].x * matrices[k + i + 1].x * matrices[j].y;
        values[k] = dpRecursion(matrices, i, k + i, cache) + dpRecursion(matrices, k + i + 1, j, cache) + m;
    }
    cache[i][j] = *min_element(values, values + (j - i));
    delete [] values;
    return cache[i][j];
}

int MatrixChainMultiplication::dynamicProgrammingRecursion(const string& filePath) {
    vector<MatrixSize> matricesInput = parseInput(filePath);
    int matricesSize = int(matricesInput.size());
    MatrixSize matrices[matricesSize];
    for (int i = 0; i < matricesSize; i++) {
        matrices[i] = matricesInput[i];
    }
    int** cache = new int*[matricesSize];
    for (int i = 0; i < matricesSize; i++) {
        cache[i] = new int[matricesSize];
        for (int j = 0; j < matricesSize; j++) {
            cache[i][j] = 0;
        }
    }
    int ans = dpRecursion(matrices, 0, matricesSize - 1, cache);
    for (int i = 0; i < matricesSize; i++) {
        delete [] cache[i];
    }
    delete [] cache;
    return ans;
}

int MatrixChainMultiplication::dynamicProgramming(const string& filePath) {
    vector<MatrixSize> matricesInput = parseInput(filePath);
    int matricesSize = int(matricesInput.size());
    MatrixSize matrices[matricesSize];
    for (int i = 0; i < matricesSize; i++) {
        matrices[i] = matricesInput[i];
    }
    int** cache = new int*[matricesSize];
    for (int i = 0; i < matricesSize; i++) {
        cache[i] = new int[matricesSize];
    }
    for (int a = 0; a < matricesSize; a++) {
        for (int i = 0; i < matricesSize - a; i++) {
            int j = a + i;
            if (j == i) {
                cache[i][j] = 0;
            } else {
                int* values = new int[j - i];
                for (int k = 0; k < j - i; k++) {
                    values[k] = cache[i][k + i] + cache[k + i + 1][j] + (matrices[i].x * matrices[k + i + 1].x * matrices[j].y);
                }
                cache[i][j] = *min_element(values, values + (j - i));
            }
        }
    }
    int ans = cache[0][matricesSize - 1];
    for (int i = 0; i < matricesSize; i++) {
        delete [] cache[i];
    }
    delete [] cache;
    return ans;
}
